django/core/validators.py

```diff
@@ -163,29 +163,27 @@ def validate_integer(value):
 
 @deconstructible
 class EmailValidator:
+    # Matches https://html.spec.whatwg.org/multipage/forms.html#valid-e-mail-address
     message = _('Enter a valid email address.')
     code = 'invalid'
     user_regex = _lazy_re_compile(
-        r"(^[-!#$%&'*+/=?^_`{}|~0-9A-Z]+(\.[-!#$%&'*+/=?^_`{}|~0-9A-Z]+)*\Z"  # dot-atom
-        r'|^"([\001-\010\013\014\016-\037!#-\[\]-\177]|\\[\001-\011\013\014\016-\177])*"\Z)',  # quoted-string
+        r"^[a-z0-9.!#$%&'*+\/=?^_`{|}~-]+\Z",
         re.IGNORECASE)
     domain_regex = _lazy_re_compile(
-        # max length for domain name labels is 63 characters per RFC 1034
-        r'((?:[A-Z0-9](?:[A-Z0-9-]{0,61}[A-Z0-9])?\.)+)(?:[A-Z0-9-]{2,63}(?<!-))\Z',
+        r'[a-z0-9](?:[a-z0-9-]{0,61}[a-z0-9])?(?:\.[a-z0-9](?:[a-z0-9-]{0,61}[a-z0-9])?)*\Z',
         re.IGNORECASE)
     literal_regex = _lazy_re_compile(
         # literal form, ipv4 or ipv6 address (SMTP 4.1.3)
         r'\[([A-f0-9:\.]+)\]\Z',
         re.IGNORECASE)
-    domain_whitelist = ['localhost']
 
     def __init__(self, message=None, code=None, whitelist=None):
         if message is not None:
             self.message = message
         if code is not None:
             self.code = code
         if whitelist is not None:
-            self.domain_whitelist = whitelist
+            pass  # deprecated
```

Comment:
I understand the deprecation policy described here https://docs.djangoproject.com/en/dev/internals/release-process/#internal-release-deprecation-policy.


Am I correct in understanding the in the current situation, a deprecation warning will be raised on Django version 2.0 and 2.1, and the actual removal of the `whitelist` attribute will happen only in version 2.2?

However, I cannot properly understand the explanation here: https://docs.djangoproject.com/en/dev/internals/contributing/writing-code/submitting-patches/#deprecating-a-feature.

I've added the deprecation behavior on `docs/releases/2.0.txt` and added a deprecation timeline to `docs/internals/deprecation.txt`. Since the release timeline here https://www.djangoproject.com/download/ suggests that the order is 2.0, 2.1 and then 2.2, I've added a `.. _deprecation-removed-in-2.2:` to the `deprecation.txt` file.

Parts that I don't understand and need help with the deprecation process is:

1. Where should `.. deprecated:: A.B` be documented? Should it look something like the following?

```
    .. attribute:: whitelist
    .. deprecated:: 2.0
        Whitelist of email domains to allow. By default, a regular expression
        (the ``domain_regex`` attribute) is used to validate whatever appears
        after the @ sign. However, if that string appears in the whitelist, this
        validation is bypassed. If not provided, the default whitelist is
        ``['localhost']``. Other domains that don't contain a dot won't pass
        validation, so you'd need to whitelist them as necessary.
```
2. What exactly are the changes that I need to make in tests? I could not find an example that implements the following:
```
from django.utils.deprecation import RemovedInDjangoXXWarning

@ignore_warnings(category=RemovedInDjangoXXWarning)
```